projects/src/simbicycle.hh

```diff
@@ -266,4 +256,59 @@ const typename Bicycle<Model, Observer>::full_state_t& Bicycle<Model, Observer>:
     return m_full_state;
 }
 
+template <typename Model, typename Observer>
+OBSERVER_FUNCTION(typename BICYCLE_TYPE::full_state_t) Bicycle<Model, Observer>::do_full_state_update(const full_state_t& full_state) {
+    // The auxiliary states _must_ also be integrated at the same time as the
+    // dynamic state. After the observer update, we "merge" dynamic states together.
+    full_state_t state_full = m_model.integrate_full_state(
+            full_state, m_input, m_model.dt(), m_measurement);
+
+    m_observer.update_state(m_input, m_measurement);
+
+    if (!m_observer.state().allFinite()) {
+        chSysHalt("");
+    }
+
+    { // limit allowed bicycle state
+        state_t x = m_observer.state();
+
+        auto limit_state_element = [&x](model::Bicycle::state_index_t index, real_t limit) {
+            const real_t value = model::Bicycle::get_state_element(x, index);
+            if (std::abs(value) > limit) {
+                model::Bicycle::set_state_element(x, index, std::copysign(limit, value));
+            }
+        };
+
+        real_t roll_angle = model_t::get_state_element(m_observer.state(),
+                model_t::state_index_t::roll_angle);
+        if (std::abs(roll_angle) > constants::pi) {
+            // state normalization limits angles to the range [-2*pi, 2*pi]
```

Comment:
Comment should read
```
// Limit roll angle to the range (-pi, pi) (exclusive)
```
https://en.wikipedia.org/wiki/Interval_(mathematics)#Notations_for_intervals